File: kakao.app/js/ked/view/menu/choices.js
var _k_ = {extend: function (c,p) {for (var k in p) { if (Object.prototype.hasOwnProperty(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}, isObj: function (o) {return !(o == null || typeof o != 'object' || o.constructor.name !== 'Object')}, empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}, max: function () { var m = -Infinity; for (var a of arguments) { if (Array.isArray(a)) {m = _k_.max.apply(_k_.max,[m].concat(a))} else {var n = parseFloat(a); if(!isNaN(n)){m = n > m ? n : m}}}; return m }, isStr: function (o) {return typeof o === 'string' || o instanceof String}, ltrim: function (s,c=' ') { while (_k_.in(s[0],c)) { s = s.slice(1) } return s}, trim: function (s,c=' ') {return _k_.ltrim(_k_.rtrim(s,c),c)}, isNum: function (o) {return !isNaN(o) && !isNaN(parseFloat(o)) && (isFinite(o) || o === Infinity || o === -Infinity)}, rtrim: function (s,c=' ') {while (_k_.in(s.slice(-1)[0],c)) { s = s.slice(0, s.length - 1) } return s}, in: function (a,l) {return (typeof l === 'string' && typeof a === 'string' && a.length ? '' : []).indexOf.call(l,a) >= 0}}

var choices

import kxk from "../../../kxk.js"
let kstr = kxk.kstr
let kseg = kxk.kseg
let slash = kxk.slash
let krzl = kxk.krzl
let post = kxk.post

import color from "../../theme/color.js"
import theme from "../../theme/theme.js"

import belt from "../../edit/tool/belt.js"

import editor from "../../edit/editor.js"


choices = (function ()
{
    _k_.extend(choices, editor)
    function choices (screen, name, features = [])
    {
        this["onKey"] = this["onKey"].bind(this)
        this["onMouse"] = this["onMouse"].bind(this)
        this["doubleClickChoiceAtIndex"] = this["doubleClickChoiceAtIndex"].bind(this)
        this["clickChoiceAtIndex"] = this["clickChoiceAtIndex"].bind(this)
        this["unhover"] = this["unhover"].bind(this)
        this["hoverChoiceAtIndex"] = this["hoverChoiceAtIndex"].bind(this)
        this["filter"] = this["filter"].bind(this)
        this["weight"] = this["weight"].bind(this)
        this["extract"] = this["extract"].bind(this)
        this["clear"] = this["clear"].bind(this)
        choices.__super__.constructor.call(this,screen,name,features)
        this.setColor('bg',theme.choices_bg)
        this.setColor('current',theme.choices_current)
        this.pointerType = 'pointer'
        this.roundedSelections = true
        this.frontRoundOffset = 0
        this.hoverIndex = -1
        this.hoverForSubmenu = false
        this.items = []
        this.fuzzied = this.items
        this.filterText = ''
    }

    choices.prototype["setColor"] = function (key, color)
    {
        choices.__super__.setColor.call(this,key,color)
    
        var _39_19_, _40_19_

        if (key === 'bg')
        {
            this.setColor('empty',this.color.bg)
            ;(this.gutter != null ? this.gutter.setColor('bg',this.color.bg) : undefined)
            return (this.scroll != null ? this.scroll.setColor('bg',this.color.bg) : undefined)
        }
    }

    choices.prototype["clear"] = function ()
    {
        return this.set([])
    }

    choices.prototype["clearEmpty"] = function ()
    {
        this.items = []
        this.fuzzied = this.items
        return this.state.clearEmpty()
    }

    choices.prototype["set"] = function (items, key)
    {
        this.items = items
        this.key = key
    
        var lines, _58_15_

        this.items = ((_58_15_=this.items) != null ? _58_15_ : [])
        this.fuzzied = this.items
        this.filterText = ''
        lines = (this.key ? this.items.map(this.extract) : this.items)
        return this.state.loadLines(lines)
    }

    choices.prototype["add"] = function (item)
    {
        this.items.push(item)
        return this.state.addLine(item.line,item.ext)
    }

    choices.prototype["append"] = function (items, ext)
    {
        this.items = this.items.concat(items)
        this.fuzzied = this.items
        return this.state.appendLines(items.map(function (i)
        {
            return i.line
        }),ext)
    }

    choices.prototype["extract"] = function (item)
    {
        return (this.key && _k_.isObj(item) ? item[this.key] : kseg.str(item))
    }

    choices.prototype["drawCursors"] = function ()
    {}

    choices.prototype["drawSelections"] = function ()
    {
        var fg, li, sel, x, xs, y

        if (_k_.empty(this.state.s.selections))
        {
            return
        }
        if (!this.roundedSelections)
        {
            return choices.__super__.drawSelections.call(this)
        }
        fg = this.color.current
        if (!this.cells.screen.t.hasFocus)
        {
            fg = color.darken(fg)
        }
        sel = this.state.s.selections[0]
        li = sel[1]
        y = li - this.state.s.view[1]
        if (y >= this.cells.rows)
        {
            return
        }
        xs = _k_.max(sel[0],kseg.headCount(this.state.s.lines[li],' '))
        if (xs === 0)
        {
            xs += this.frontRoundOffset
        }
        this.cells.set_ch_fg(xs - 1 - this.state.s.view[0],y,'',fg)
        for (var _a_ = x = xs, _b_ = sel[2]; (_a_ <= _b_ ? x < sel[2] : x > sel[2]); (_a_ <= _b_ ? ++x : --x))
        {
            this.cells.set_bg(x - this.state.s.view[0],y,fg)
            this.cells.adjustContrastForHighlight(x - this.state.s.view[0],y,fg)
        }
        return this.cells.set_ch_fg(x - this.state.s.view[0],y,'',fg)
    }

    choices.prototype["numChoices"] = function ()
    {
        return this.items.length
    }

    choices.prototype["numFiltered"] = function ()
    {
        return this.fuzzied.length
    }

    choices.prototype["currentIndex"] = function ()
    {
        return this.state.mainCursor()[1]
    }

    choices.prototype["current"] = function (opt)
    {
        var cc

        cc = this.fuzzied[this.currentIndex()]
        if (_k_.isStr(cc))
        {
            if ((opt != null ? opt.trim : undefined) === 'front')
            {
                cc = _k_.ltrim(cc)
            }
            else if ((opt != null ? opt.trim : undefined) !== false)
            {
                cc = _k_.trim(cc)
            }
        }
        return cc
    }

    choices.prototype["choiceAtRow"] = function (row)
    {
        return this.fuzzied[row]
    }

    choices.prototype["hasNext"] = function ()
    {
        var _137_26_

        return (this.nextRow() != null)
    }

    choices.prototype["hasPrev"] = function ()
    {
        var _138_26_

        return (this.prevRow() != null)
    }

    choices.prototype["nextRow"] = function ()
    {
        var y

        y = this.state.mainCursor()[1]
        while (y < this.state.s.lines.length - 1)
        {
            y += 1
            if (kseg.trim(this.state.s.lines[y]).length)
            {
                return y
            }
        }
    }

    choices.prototype["prevRow"] = function ()
    {
        var y

        y = this.state.mainCursor()[1]
        while (y > 0)
        {
            y -= 1
            if (kseg.trim(this.state.s.lines[y]).length)
            {
                return y
            }
        }
    }

    choices.prototype["select"] = function (row)
    {
        if (!(_k_.isNum(row)))
        {
            return
        }
        if (row < 0 || row >= this.state.s.lines.length)
        {
            return
        }
        this.state.setSelections([belt.rangeOfLine(this.state.s.lines,row)])
        this.state.setMainCursor(0,row)
        if (this.focusable)
        {
            this.grabFocus()
        }
        return this.emit('select',this.choiceAtRow(row))
    }

    choices.prototype["selectFirst"] = function ()
    {
        return this.select(0)
    }

    choices.prototype["moveSelection"] = function (dir)
    {
        switch (dir)
        {
            case 'down':
                this.selectNext()
                break
            case 'up':
                this.selectPrev()
                break
        }

        return this
    }

    choices.prototype["selectNext"] = function ()
    {
        return this.select(this.nextRow())
    }

    choices.prototype["selectPrev"] = function ()
    {
        return this.select(this.prevRow())
    }

    choices.prototype["weight"] = function (item, text)
    {
        var itemText, matchOrLevenshtein, p, w

        itemText = this.extract(item)
        p = slash.parse(itemText)
        matchOrLevenshtein = function (t)
        {
            var idx

            idx = t.indexOf(text)
            if (idx < 0)
            {
                idx = t.length + kstr.levensthein(t,text)
            }
            return idx
        }
        w = this.items.indexOf(item)
        w += 10 * matchOrLevenshtein(p.name)
        w += 5 * matchOrLevenshtein(p.dir)
        w += (!_k_.empty(p.ext) ? (0.1 * matchOrLevenshtein(p.ext)) : 4)
        return w
    }

    choices.prototype["filter"] = function (text)
    {
        var fuzz, lines

        if (_k_.empty(this.items))
        {
            return
        }
        if (text === this.filterText)
        {
            return
        }
        if (_k_.empty(text))
        {
            return this.set(this.items,this.key)
        }
        this.filterText = text
        fuzz = new krzl({values:this.items,extract:this.extract})
        this.fuzzied = fuzz.filter(text)
        this.fuzzied.sort((function (a, b)
        {
            return this.weight(a,text) - this.weight(b,text)
        }).bind(this))
        lines = this.fuzzied.map(this.extract)
        if (_k_.empty(lines))
        {
            lines = ['']
        }
        return this.state.loadLines(lines)
    }

    choices.prototype["hoverChoiceAtIndex"] = function (index, event)
    {
        if (this.hoverIndex === index)
        {
            return true
        }
        this.hoverIndex = index
        this.state.setMainCursor(0,index)
        this.select(this.hoverIndex)
        post.emit('pointer','pointer')
        this.emitAction('hover',this.current(),event)
        return {redraw:true}
    }

    choices.prototype["unhover"] = function ()
    {
        this.hoverIndex = -1
        this.state.deselect()
        post.emit('pointer','default')
        return {redraw:true}
    }

    choices.prototype["clickChoiceAtIndex"] = function (index, event)
    {
        this.hoverIndex = -1
        this.emitAction('click',this.fuzzied[index],event)
        return {redraw:true}
    }

    choices.prototype["doubleClickChoiceAtIndex"] = function (index, event)
    {
        this.hoverIndex = -1
        this.emitAction('doubleclick',this.fuzzied[index],event)
        return {redraw:true}
    }

    choices.prototype["onMouse"] = function (event)
    {
        var col, dx, dy, ret, row, _275_21_

        ret = choices.__super__.onMouse.call(this,event)
        if ((ret != null ? ret.redraw : undefined))
        {
            return ret
        }
        ret = (this.mapscr != null ? this.mapscr.onMouse(event) : undefined)
        if (ret)
        {
            return ret
        }
        if (this.hover)
        {
            var _a_ = this.cells.posForEvent(event); col = _a_[0]; row = _a_[1]

            if (this.state.isValidLineIndex(row))
            {
                if (this.hoverForSubmenu && event.type === 'move' && col > kseg.width(this.state.s.lines[row]))
                {
                    dx = Math.abs(event.delta[0])
                    dy = Math.abs(event.delta[1])
                    if (dx * 2 >= dy)
                    {
                        return
                    }
                }
                switch (event.type)
                {
                    case 'move':
                        return this.hoverChoiceAtIndex(row + this.state.s.view[1],event)

                    case 'press':
                        if (event.count === 2)
                        {
                            return this.doubleClickChoiceAtIndex(row + this.state.s.view[1],event)
                        }
                        else
                        {
                            return this.clickChoiceAtIndex(row + this.state.s.view[1],event)
                        }
                        break
                }

            }
        }
        return this.hover
    }

    choices.prototype["emitAction"] = function (action, arg, event)
    {
        return this.emit('action',action,arg,event)
    }

    choices.prototype["onKey"] = function (key, event)
    {
        switch (event.combo)
        {
            case 'up':
            case 'down':
                this.moveSelection(event.combo)
                return true

        }

        if (!this.hasFocus())
        {
            return
        }
        switch (event.combo)
        {
            case 'esc':
            case 'left':
            case 'right':
            case 'space':
            case 'delete':
            case 'return':
                this.emitAction(event.combo,this.current(),event)
                break
        }

        return true
    }

    return choices
})()

global.choices_class = choices
export default choices;

File: kakao.app/love/lua/view/menu/choices.lua
--[[
     ███████  ███   ███   ███████   ███   ███████  ████████   ███████  
    ███       ███   ███  ███   ███  ███  ███       ███       ███       
    ███       █████████  ███   ███  ███  ███       ███████   ███████   
    ███       ███   ███  ███   ███  ███  ███       ███            ███  
     ███████  ███   ███   ███████   ███   ███████  ████████  ███████   

    an editor that can't edit ;)
    
    displays selections with nice rounded edges and is used almost 
    everywhere by now: 
        context menu dirtree funtree quicky browse finder searcher...
    
    takes list of items as input and extracts text to display from them    
    supports filtering of the items
--]]

editor = require "edit.editor"


local choices = class("choices", editor)
    


function choices:init(name, features) 
        features = features or (array())
        
        editor.init(self, name, features)
        
        self:setColor('bg', theme.editor.bg)
        self:setColor('hover', theme.hover)
        
        self.pointerType = 'pointer'
        self.roundedSelections = true
        self.frontRoundOffset = 0
        self.hoverIndex = -1
        self.hoverForSubmenu = false
        
        self.items = array()
        self.fuzzied = self.items
        self.filterText = ''
        return self
    end


function choices:setColor(key, color) 
        editor.setColor(self, key, color)
        
        if (key == 'bg') then 
            self:setColor('empty', self.color.bg)
            if self.gutter then 
                self.gutter:setColor('bg', self.color.bg)
            end
            
            if self.scroll then 
                return self.scroll:setColor('bg', self.color.bg)
            end
        end
    end


function choices:extract(item) 
        if self.key then 
            return item[self.key]
        else 
            return kseg.str(item)
        end
    end


function choices:clear() 
    return self:set(array())
    end


function choices:clearEmpty() 
        self.items = array()
        self.fuzzied = self.items
        return self.state:clearEmpty()
    end

--  0000000  00000000  000000000  
-- 000       000          000     
-- 0000000   0000000      000     
--      000  000          000     
-- 0000000   00000000     000     


function choices:set(items, key) 
        self.key = key
        self.items = items
        self.items = self.items or (array())
        self.fuzzied = self.items
        self.filterText = ''
        
        local lines = self.items
        -- log "CHOICES #{@name} set lines #{key} " lines
        if self.key then lines = self.items:map(function (i) return self:extract(i) end) end
        -- log "CHOICES #{@name} set lines #{key} " lines
        self.state:loadLines(lines)
        return self
    end


function choices:add(item) 
        self.items:push(item)
        return self.state:addLine(item.line, item.ext)
    end


function choices:append(items, ext, key) 
        key = key or 'line'
        
        self.items = self.items.concat(items)
        self.fuzzied = self.items -- 𝜏𝖍𝚒𝖘 𝚒𝖘 ⟒ɼ⊚∩𝚐! 𝚒𝜏ϵ⫙𝖘 𝖘𝖍⊚𝓊⟅𝒹 𝔟ϵ ɼϵϝ𝓊𝓏𝓏𝚒ϵ𝒹 𝚒∩𝖘𝜏ϵ𝕒𝒹
        return self.state:appendLines(items:map(function (i) return i[key] end), ext)
    end

-- 0000000    00000000    0000000   000   000  
-- 000   000  000   000  000   000  000 0 000  
-- 000   000  0000000    000000000  000000000  
-- 000   000  000   000  000   000  000   000  
-- 0000000    000   000  000   000  00     00  


function choices:drawCursors() 
    
    end


function choices:drawSelections() 
        if empty(self.state.s.selections) then return end
        
        if not self.roundedSelections then 
            return editor.drawSelections(self)
        end
        
        -- fg = if @hasFocus() ➜ @color.hover.bg ➜ @color.hover.blur
        local fg = self.color.hover.bg
        
        -- if not @cells.screen.t.hasFocus
        --     fg = color.darken fg
        
        local sel = self.state.s.selections[1]
        
        if not sel then return end
        
        local li = sel[2]
        local y = ((li - self.state.s.view[2]) + 1)
        
        if (y > self.cells.rows) then return end
        
        local xs = max(sel[1], kseg.headCount(self.state.s.lines[li], ' '))
        if (xs == 1) then 
            xs = xs + (self.frontRoundOffset)
        end
        
        self.cells:set_ch_fg(xs, y, '', fg)
        
        for x in iter((xs + 1), sel[3]) do 
            self.cells:set_bg(((x - self.state.s.view[1]) + 1), y, fg)
            self.cells:adjustContrastForHighlight(((x - self.state.s.view[1]) + 1), y, fg)
        end
        
        return self.cells:set_ch_fg(((sel[3] + 2) - self.state.s.view[1]), y, '', fg)
    end


function choices:numChoices() 
    return #self.items
    end

function choices:numFiltered() 
    return #self.fuzzied
    end


function choices:currentIndex() 
    return self.state:mainCursor()[2]
    end

function choices:current(opt) 
        opt = opt or ({})
        local cc = self.fuzzied[self:currentIndex()]
        if is(cc, "string") then 
            if (opt.trim == 'front') then 
                cc = kstr.ltrim(cc)
            elseif (opt.trim ~= false) then 
                cc = kstr.trim(cc)
            end
        end
        
        return cc
    end


function choices:choiceAtRow(row) 
    return self.fuzzied[row]
    end

-- 000   000  00000000  000   000  000000000        00000000   00000000   00000000  000   000  
-- 0000  000  000        000 000      000           000   000  000   000  000       000   000  
-- 000 0 000  0000000     00000       000           00000000   0000000    0000000    000 000   
-- 000  0000  000        000 000      000           000        000   000  000          000     
-- 000   000  00000000  000   000     000           000        000   000  00000000      0      


function choices:hasNext() 
    return self:nextRow()
    end

function choices:hasPrev() 
    return self:prevRow()
    end


function choices:nextRow() 
        local y = self:currentIndex()
        while (y < self.state.s.lines:len()) do 
            y = y + 1
            if (kseg.trim(self.state.s.lines[y]):len() >= 1) then 
                return y
            end
        end
    end


function choices:prevRow() 
        local y = self:currentIndex()
        while (y > 1) do 
            y = y - 1
            if (kseg.trim(self.state.s.lines[y]):len() >= 1) then 
                return y
            end
        end
    end


function choices:pageUpRow() 
        local y = ((self:currentIndex() - self.cells.rows) + 1)
        y = max(y, 1)
        
        while ((y > 1) and empty(kseg.trim(self.state.s.lines[y]))) do 
            y = y - 1
        end
        
        if empty((kseg.trim(self.state.s.lines[y]) and (y < self.state.s.lines:len()))) then 
            y = y + 1
        end
        
        return y
    end


function choices:pageDownRow() 
        local y = ((self:currentIndex() + self.cells.rows) - 1)
        y = min(y, self.state.s.lines:len())
        
        while ((y < self.state.s.lines:len()) and empty(kseg.trim(self.state.s.lines[y]))) do 
            y = y + 1
        end
        
        if empty((kseg.trim(self.state.s.lines[y]) and (y > 1))) then 
            y = y - 1
        end
        
        return y
    end

--  0000000  00000000  000      00000000   0000000  000000000  
-- 000       000       000      000       000          000     
-- 0000000   0000000   000      0000000   000          000     
--      000  000       000      000       000          000     
-- 0000000   00000000  0000000  00000000   0000000     000     


function choices:select(row) 
        if not row then return end
        if is(not row, "number") then return end
        if ((row < 1) or (row > self.state.s.lines:len())) then return end
        
        self.state:setSelections(array(belt.rangeOfLine(self.state.s.lines, row)))
        self.state:setMainCursor(1, row)
        
        if self.focusable then self:grabFocus() end
        
        return self:emit('select', self:choiceAtRow(row))
    end


function choices:selectFirst() 
    return self:select(1)
    end


function choices:moveSelection(dir) 
        if (dir == 'pagedown') then self:selectPageDown()
        elseif (dir == 'pageup') then self:selectPageUp()
        elseif (dir == 'down') then self:selectNext()
        elseif (dir == 'up') then self:selectPrev()
        end
        
        return self
    end


function choices:selectPageUp() 
    return self:select(self:pageUpRow())
    end

function choices:selectPageDown() 
    return self:select(self:pageDownRow())
    end

function choices:selectNext() 
    return self:select(self:nextRow())
    end

function choices:selectPrev() 
        local row = self:prevRow()
        if empty(row) then 
            return self:emitAction('boundary', self:current())
        else 
            return self:select(row)
        end
    end

-- 000   000  00000000  000   0000000   000   000  000000000  
-- 000 0 000  000       000  000        000   000     000     
-- 000000000  0000000   000  000  0000  000000000     000     
-- 000   000  000       000  000   000  000   000     000     
-- 00     00  00000000  000   0000000   000   000     000     


function choices:weight(item, text) 
        local itemText = self:extract(item)
        local p = slash.parse(itemText)
        
        
        function matchOrLevenshtein(t) 
            local idx = t.indexOf(text)
            if (idx < 0) then 
                idx = (#t + kstr.levensthein(t, text))
            end
            
            return idx
        end
        
        local w = self.items.indexOf(item) -- try to keep order of original items
        w = w + ((10 * matchOrLevenshtein(p.name))) -- high focus on file name
        w = w + ((5 * matchOrLevenshtein(p.dir))) -- lesser focus on dir
        -- w += valid p.ext ? (0.1 * matchOrLevenshtein(p.ext)) : 4 # low weight for extensions
        if valid(p.ext) then 
            w = w + ((0.1 * matchOrLevenshtein(p.ext)))
        else 
            w = w + 4
        end
        
        return w
    end

-- 00000000  000  000      000000000  00000000  00000000   
-- 000       000  000         000     000       000   000  
-- 000000    000  000         000     0000000   0000000    
-- 000       000  000         000     000       000   000  
-- 000       000  0000000     000     00000000  000   000  


function choices:filter(text) 
        if empty(self.items) then return end
        
        if (text == self.filterText) then return end
        
        if empty(text) then return self:set(self.items, self.key) end
        
        self.filterText = text
        
        local fuzz = krzl({values = self.items, extract = self.extract})
        
        self.fuzzied = fuzz:filter(text)
        
        self.fuzzied:sort(function (a, b) 
    return (self:weight(a, text) < self:weight(b, text))
end)
        
        local lines = self.fuzzied.map(self.extract)
        
        if empty(lines) then 
            lines = array('')
        end
        
        return self.state:loadLines(lines)
    end

-- ███   ███   ███████   ███   ███  ████████  ████████ 
-- ███   ███  ███   ███  ███   ███  ███       ███   ███
-- █████████  ███   ███   ███ ███   ███████   ███████  
-- ███   ███  ███   ███     ███     ███       ███   ███
-- ███   ███   ███████       █      ████████  ███   ███


function choices:hoverChoiceAtIndex(index, event) 
        if (self.hoverIndex == index) then return true end
        
        self.hoverIndex = index
        self.state:setMainCursor(0, index)
        self:select(self.hoverIndex)
        post:emit('pointer', 'pointer')
        self:emitAction('hover', self:current(), event)
        return true
    end


function choices:unhover() 
        self.hoverIndex = -1
        self.state:deselect()
        post:emit('pointer', 'default')
        return true
    end

-- 00     00   0000000   000   000   0000000  00000000  
-- 000   000  000   000  000   000  000       000       
-- 000000000  000   000  000   000  0000000   0000000   
-- 000 0 000  000   000  000   000       000  000       
-- 000   000   0000000    0000000   0000000   00000000  


function choices:dragChoiceAtIndex(index, event) 
        self:hoverChoiceAtIndex(index, event)
        self:emitAction('drag', self.fuzzied[index], event)
        return {redraw = true}
    end


function choices:clickChoiceAtIndex(index, event) 
        self.hoverIndex = -1
        self:emitAction('click', self.fuzzied[index], event)
        return true
    end


function choices:doubleClickChoiceAtIndex(index, event) 
        self.hoverIndex = -1
        self:emitAction('doubleclick', self.fuzzied[index], event)
        return true
    end


function choices:onMouse(event) 
        if editor.onMouse(self, event) then return true end
        
        if (self.mapscr and self.mapscr.onMouse(event)) then return true end
        
        if self.hover then 
            local col, row = unpack(self:eventPos(event))
            
            if self.state:isValidLineIndex(row) then 
                if ((self.hoverForSubmenu and (event.type == 'move')) and (col > kseg.width(self.state.s.lines[row]))) then 
                    local dx = abs(event.delta[1])
                    local dy = abs(event.delta[2])
                    if ((dx * 2) >= dy) then return end
                end
                
                local index = ((row + self.state.s.view[2]) - 1)
                
                if (event.type == 'move') then 
                            return self:hoverChoiceAtIndex(index, event)
                elseif (event.type == 'press') then 
                        if (event.count == 2) then 
                            return self:doubleClickChoiceAtIndex(index, event)
                        else 
                            self.dragStart = array(col, row)
                            return self:clickChoiceAtIndex(index, event)
                        end
                elseif (event.type == 'drag') then 
                            if self.dragStart then 
                                return self:dragChoiceAtIndex(index, event)
                            end
                elseif (event.type == 'release') then 
                            self.dragStart = nil
                end
            end
        end
        
        return self.hover
    end

--  0000000    0000000  000000000  000   0000000   000   000  
-- 000   000  000          000     000  000   000  0000  000  
-- 000000000  000          000     000  000   000  000 0 000  
-- 000   000  000          000     000  000   000  000  0000  
-- 000   000   0000000     000     000   0000000   000   000  


function choices:emitAction(action, choice, event) 
        -- log "emitAction" action, choice, event
        return self:emit('action', action, choice, event)
    end

-- 000   000  00000000  000   000  
-- 000  000   000        000 000   
-- 0000000    0000000     00000    
-- 000  000   000          000     
-- 000   000  00000000     000     


function choices:onKey(key, event) 
        -- log "CHOICES EMIT ACTION #{@name} onKey #{key} combo #{event.combo}"
        
        if (event.combo == 'up') or (event.combo == 'down') or (event.combo == 'pageup') or (event.combo == 'pagedown') then 
                self:moveSelection(event.combo)
                return true
        elseif (event.combo == 'ctrl+alt+up') then self:moveSelection('pageup') ; return true
        elseif (event.combo == 'ctrl+alt+down') then self:moveSelection('pagedown') ; return true
        end
        
        if not self:hasFocus() then return end
        
        if (event.combo == 'esc') or (event.combo == 'left') or (event.combo == 'right') or (event.combo == 'space') or (event.combo == 'delete') or (event.combo == 'return') then self:emitAction(event.combo, self:current(), event)
        end
        
        -- not calling super here effectively disables all text-editing
        return true
    end

-- _G.choices_class = choices # hack to prevent import recursion
return choices